File: MobileMeteo/MMTCitiesStore.m
//
//  MMTCitiesStore.m
//  Mobile Meteo
//
//  Created by Kamil Szostakowski on 08.05.2015.
//  Copyright (c) 2015 Kamil Szostakowski. All rights reserved.
//

#import "MMTCitiesStore.h"
#import <CoreLocation/CoreLocation.h>

@interface MMTCitiesStore ()

@property (nonatomic, strong) CLGeocoder *geocoder;
@end

@implementation MMTCitiesStore

- (NSArray*)getAllCities
{
    NSArray *cities = @[];
    NSArray *names = @[@"Białystok", @"Bydgoszcz", @"Gdańsk", @"Gorzów Wielkopolski", @"Katowice", @"Kielce", @"Kraków", @"Lublin", @"Łódź",
             @"Olsztyn", @"Opole", @"Poznań", @"Rzeszów", @"Szczecin", @"Toruń", @"Warszawa", @"Wrocław", @"Zielona Góra"];
    
    for (NSString *name in names)
    {
        cities = [cities arrayByAddingObject:[self cityWithName:name region:@"" location:nil]];
    }
    
    return cities;
}

- (NSArray*)getCitiesMachingCriteria:(NSString*)criteria
{
    return [[self getAllCities] filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"SELF.name CONTAINS[cd] %@", criteria]];
}

- (void)findCitiesMachingCriteris:(NSString*)criteria completion:(ListFetchCompletion)completion
{
    if(!self.geocoder)
    {
        self.geocoder = [CLGeocoder new];
    }
    
    NSString *addressString = [NSString stringWithFormat:@"%@, Poland", criteria];
    
    [self.geocoder geocodeAddressString:addressString completionHandler:^(NSArray *placemarks, NSError *error)
    {
        NSArray *cities = @[];
        
        for (CLPlacemark *placemark in placemarks)
        {
            cities = [cities arrayByAddingObject:[self cityWithName:placemark.name region:placemark.administrativeArea location:placemark.location]];
        }                        
        
        completion(cities);
    }];
}

#pragma mark - Helper methods

- (NSDictionary*)cityWithName:(NSString*)name region:(NSString*)region location:(CLLocation*)location
{
    return @
    {
        @"name": name,
        @"region": region,
        @"location": location ? location : [NSNull null],
    };
}

@end

File: MobileMeteo/MMTCitiesStore.swift
//
//  MMTCitiesStore.swift
//  MobileMeteo
//
//  Created by Kamil Szostakowski on 07.07.2015.
//  Copyright (c) 2015 Kamil Szostakowski. All rights reserved.
//

import Foundation
import CoreLocation
import AddressBook
import CoreData

typealias MMTCityQueryCompletion = (MMTCityProt?, MMTError?) -> Void
typealias MMTCitiesQueryCompletion = ([MMTCityProt]) -> Void

class MMTCitiesStore
{
    private typealias MMTCurrentCityQueryCompletion = (MMTCityProt?) -> Void
    
    private let database: MMTDatabase
    private let geocoder: MMTCityGeocoder
    
    init(db: MMTDatabase, geocoder gcoder: MMTCityGeocoder)
    {
        geocoder = gcoder
        database = db                
    }
    
    // MARK: Methods    
    
    func getAllCities(_ completion: MMTCitiesQueryCompletion)
    {
        completion(getAllCities())
    }
    
    func findCityForLocation(_ location: CLLocation, completion: @escaping MMTCityQueryCompletion)
    {
        geocoder.city(for: location) {
            (geocodedCity: MMTCityProt?, error: MMTError?) in
            
            var city = geocodedCity;
            defer { completion(city, error) }
            
            guard error == nil else { return }
            guard let localCity = self.getCity(with: city!.name) else { return; }
            
            city = localCity
        }
    }
    
    func findCitiesMatchingCriteria(_ criteria: String, completion: @escaping MMTCitiesQueryCompletion)
    {
        let cities = getCitiesMatchingCriteria(criteria)
        
        if cities.count > 0 { completion(cities)
            return
        }
        
        geocoder.cities(matching: criteria, completion: completion)
    }
    
    func markCity(_ city: MMTCityProt, asFavourite favourite: Bool)
    {
        guard let aCity = city as? MMTCity else {
            return
        }
        
        city.isFavourite = favourite

        if favourite && aCity.managedObjectContext == nil {
            database.context.insert(aCity)
        }
        
        if favourite == false && city.isCapital == false {
            database.context.delete(aCity)
        }
        
        database.saveContext()
    }        
    
    // MARK: Helper methods
    
    private func getAllCities() -> [MMTCityProt]
    {
        let fetchRequest = citiesFetchRequest(predicate: nil)
        return (try? database.context.fetch(fetchRequest)) ?? []
    }
    
    private func getCitiesMatchingCriteria(_ criteria: String) -> [MMTCityProt]
    {
        let fetchRequest = citiesFetchRequest(predicate: NSPredicate(format: "SELF.name CONTAINS[cd] %@", criteria))
        return (try? database.context.fetch(fetchRequest)) ?? []
    }
    
    private func getCity(with name: String) -> MMTCityProt?
    {
        let
        fetchRequest = citiesFetchRequest(predicate: NSPredicate(format: "SELF.name LIKE[cd] %@", name))
        fetchRequest.fetchLimit = 1
        
        return (try? database.context.fetch(fetchRequest))?.first
    }
    
    private func citiesFetchRequest(predicate: NSPredicate?) -> NSFetchRequest<MMTCity>
    {
        let
        fetchRequest = NSFetchRequest<MMTCity>(entityName: "MMTCity")
        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "name", ascending: true)]
        fetchRequest.predicate = predicate
        
        return fetchRequest
    }
}
